eth/downloader/beaconsync.go

```diff
@@ -288,6 +289,48 @@ func (d *Downloader) fetchBeaconHeaders(from uint64) error {
 		log.Warn("Retrieved beacon headers from local", "from", from, "count", count)
 	}
 	for {
+		// Some beacon headers might have appeared since the last cycle, make
+		// sure we're always syncing to all available ones
+		head, _, err = d.skeleton.Bounds()
+		if err != nil {
+			return err
+		}
+		// If the pivot became stale (older than 2*64-8 (bit of wiggle room)),
+		// move it ahead to HEAD-64
+		d.pivotLock.Lock()
+		if d.pivotHeader != nil {
+			if head.Number.Uint64() > d.pivotHeader.Number.Uint64()+2*uint64(fsMinFullBlocks)-8 {
+				// Retrieve the next pivot header, either from skeleton chain
+				// or the filled chain
+				number := head.Number.Uint64() - uint64(fsMinFullBlocks)
+
+				log.Warn("Pivot seemingly stale, moving", "old", d.pivotHeader.Number, "new", number)
+				if d.pivotHeader = d.skeleton.Header(number); d.pivotHeader == nil {
```

Comment:
Actually it's possible to happen(new pivot point is still in local)

For example:

- current remote head is 101
- current local head is 99
- skeleton header chain range is [100, 101]
- current pivot point is `101-64=37`
- start sync
- new remote head is 158
- `158-37 > 2*64-8`
- new pivot point is 94
- new pivot point is still in range of local chain

If I leave some nitpicks, is that the local chain part is already resolved previously https://github.com/ethereum/go-ethereum/pull/26453/files#diff-8cb29165929b2bd4c8f7c019a95cc8a69d3e36320703308f5c835fa97bdd8f31R288

But I think it's fine. Just a very very special corner case.

Reply:
@rjl493456442 do you have any suggestion on a change here?

Reply:
pushed